File: packages/julia/src/Seguranca/Ataques/AtaqueFoldEstrutural.jl
"""
Esse é o ataque que pegou o bug mais sério que encontramos: quando a
distância de mistura era exatamente metade do bloco, TODOS os IVs
aleatórios produziam um bloco cuja primeira metade era idêntica à segunda.
Testa repetição em frações de 1/2, 1/4 e 1/8 do bloco.
"""
struct AtaqueFoldEstrutural <: Ataque
    amostras::Int
    tamanho_bloco::Int
end

AtaqueFoldEstrutural(; amostras::Integer = 5000, tamanho_bloco::Integer = 32) =
    AtaqueFoldEstrutural(Int(amostras), Int(tamanho_bloco))

nome(::AtaqueFoldEstrutural) = "Fold estrutural (metades/quartos/oitavos repetidos)"

function executar(a::AtaqueFoldEstrutural, alvo::AlvoCriptografico)
    chave = chave_de_teste(alvo)

    divisores = [
        d for d in [2, 4, 8, 16] if a.tamanho_bloco % d == 0 && div(a.tamanho_bloco, d) >= 1
    ]
    ocorrencias = Dict{Int,Int}(d => 0 for d in divisores)

    # Limite de "ruído esperado" por acaso: pra fatias de $tam bytes,
    # a chance de colisão por acaso é ~1/256^tam por par comparado -
    # desprezível para tam >= 2, então qualquer contagem > 0 já é
    # suspeita o bastante pra investigar (ajustamos a margem pra
    # fatias de 1-2 bytes, onde colisão por acaso é mais provável).
    for _ in 1:(a.amostras)
        ks = gerar_keystream_bruto(
            alvo,
            chave,
            random_bytes(tamanho_iv(alvo)),
            "enc",
            a.tamanho_bloco,
        )
        for divisor in divisores
            tam_fatia = div(a.tamanho_bloco, divisor)
            primeira_fatia = ks[1:tam_fatia]
            for j in 1:(divisor-1)
                if ks[j*tam_fatia+1:(j+1)*tam_fatia] == primeira_fatia
                    ocorrencias[divisor] += 1
                    break
                end
            end
        end
    end

    margem(divisor::Integer) =
        div(a.tamanho_bloco, divisor) <= 2 ? trunc(Int, a.amostras * 0.01) + 3 : 5

    problemas = Dict{Int,Int}()
    for divisor in divisores
        c = ocorrencias[divisor]
        if c > margem(divisor)
            problemas[divisor] = c
        end
    end

    if !isempty(problemas)
        detalhe = join(
            [
                "1/$d do bloco repetido em $(problemas[d])/$(a.amostras)"
                for d in divisores if haskey(problemas, d)
            ],
            ", ",
        )
        return ResultadoAtaque(
            nome(a),
            true,
            "critica",
            detalhe,
            Dict{String,Any}("ocorrencias" => ocorrencias),
        )
    end

    json_ocorrencias =
        "{" * join(["\"$d\":$(ocorrencias[d])" for d in divisores], ",") * "}"

    return ResultadoAtaque(
        nome(a),
        false,
        "info",
        "Nenhuma repetição estrutural acima do ruído esperado: " * json_ocorrencias,
    )
end

File: packages/rust/src/seguranca/ataques/ataque_fold_estrutural.rs
//! Esse é o ataque que pegou o bug mais sério que encontramos: quando a
//! distância de mistura era exatamente metade do bloco, TODOS os IVs
//! aleatórios produziam um bloco cuja primeira metade era idêntica à segunda.
//! Testa repetição em frações de 1/2, 1/4 e 1/8 do bloco.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::random_bytes;

pub struct AtaqueFoldEstrutural {
    pub amostras: usize,
    pub tamanho_bloco: usize,
}

impl AtaqueFoldEstrutural {
    pub fn new() -> Self {
        AtaqueFoldEstrutural {
            amostras: 5000,
            tamanho_bloco: 32,
        }
    }
}

impl Default for AtaqueFoldEstrutural {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueFoldEstrutural {
    fn nome(&self) -> String {
        "Fold estrutural (metades/quartos/oitavos repetidos)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let chave = alvo.chave_de_teste();

        let divisores: Vec<usize> = [2usize, 4, 8, 16]
            .into_iter()
            .filter(|&d| self.tamanho_bloco.is_multiple_of(d) && self.tamanho_bloco / d >= 1)
            .collect();

        let mut ocorrencias: Vec<(usize, usize)> =
            divisores.iter().map(|&d| (d, 0usize)).collect();

        // Limite de "ruído esperado" por acaso: pra fatias de $tam bytes,
        // a chance de colisão por acaso é ~1/256^tam por par comparado -
        // desprezível para tam >= 2, então qualquer contagem > 0 já é
        // suspeita o bastante pra investigar (ajustamos a margem pra
        // fatias de 1-2 bytes, onde colisão por acaso é mais provável).
        for _ in 0..self.amostras {
            let ks = alvo.gerar_keystream_bruto(
                chave.as_bytes(),
                &random_bytes(alvo.tamanho_iv()),
                "enc",
                self.tamanho_bloco,
            );
            for divisor in &divisores {
                let tam_fatia = self.tamanho_bloco / divisor;
                let primeira_fatia = &ks[0..tam_fatia];
                for j in 1..*divisor {
                    if &ks[j * tam_fatia..j * tam_fatia + tam_fatia] == primeira_fatia {
                        if let Some(entry) = ocorrencias.iter_mut().find(|(d, _)| d == divisor) {
                            entry.1 += 1;
                        }
                        break;
                    }
                }
            }
        }

        let margem = |divisor: usize| -> usize {
            if self.tamanho_bloco / divisor <= 2 {
                (self.amostras as f64 * 0.01) as usize + 3
            } else {
                5
            }
        };

        let problemas: Vec<(usize, usize)> = ocorrencias
            .iter()
            .filter(|(d, c)| *c > margem(*d))
            .cloned()
            .collect();

        if !problemas.is_empty() {
            let detalhe = problemas
                .iter()
                .map(|(d, c)| format!("1/{} do bloco repetido em {}/{}", d, c, self.amostras))
                .collect::<Vec<String>>()
                .join(", ");
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Critica,
                &detalhe,
            ));
        }

        let json: Vec<String> = ocorrencias
            .iter()
            .map(|(d, c)| format!("\"{}\":{}", d, c))
            .collect();
        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &("Nenhuma repetição estrutural acima do ruído esperado: {".to_string()
                + &json.join(",")
                + "}"),
        ))
    }
}
